segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeperstorage/LogStorageManager.java

```diff
@@ -0,0 +1,741 @@
+            this.sealLedger(ledger.getLastLedgerData());
+        } catch (Exception exc) {
+            throw translateZKException(segmentName, exc);
+        }
+    }
+
+    /**
+     * Concatenates the sourceSegment in to the target segment at offset.
+     *
+     * The concatenation involves updating the metadata in a single ZK transaction.
+     * The operations are:
+     * 1. Add the ledgers of the source segment to the target metadata at their new offset.
+     * 2. Remove them from the source metadata.
+     * 3. Update the target version to ensure CAS.
+     *
+     * @param segmentName the target segment.
+     * @param sourceSegment The segment to be merged to target.
+     * @param offset offset at which the merge happens.
+     * @return A completable future which completes once the concat operation is complete.
+     */
+    public void concat(String segmentName, String sourceSegment, long offset) throws StreamSegmentException {
+        List<CuratorOp> curatorOps;
+        try {
+        curatorOps = this.getZKOperationsForConcat(segmentName, sourceSegment, offset);
+            List<CuratorTransactionResult> results = zkClient.transaction()
+                                                             .forOperations(curatorOps);
+            LogStorage logStorage = getOrRetrieveStorageLedger(segmentName);
+            logStorage.incrementUpdateVersion();
+            /* Fence all the ledgers and add one at the end. */
+            fenceLedgersAndCreateOneAtEnd(segmentName, logStorage);
+            /** Delete metadata and cache for the source once the operation is complete. */
+            this.zkClient.delete().deletingChildrenIfNeeded().forPath(getZkPath(sourceSegment));
+            ledgers.remove(sourceSegment);
+        } catch (Exception exc) {
+            /** In case of exception, drop the corrupt cache. */
+            ledgers.remove(segmentName);
+            throw translateZKException(segmentName, exc);
+        }
+    }
+
+    /**
+     * Deletes a segment along with all the data and metadata.
+     * @param segmentName name of the segment to be deleted.
+     * @return A CompletableFuture which completes once the delete is complete.
+     */
+    public void delete(String segmentName) throws StreamSegmentException {
+        LogStorage ledger = this.getOrRetrieveStorageLedger(segmentName);
+        if (ledger.getContainerEpoch() > this.containerEpoch.get()) {
+            throw new CompletionException(new StorageNotPrimaryException(segmentName));
+        }
+        try {
+            this.zkClient.delete()
+                         .deletingChildrenIfNeeded()
+                         .forPath(getZkPath(segmentName));
+        } catch (Exception e) {
+           log.warn("Exception while deleting a segment {}", segmentName, e);
+
+        }
+        ledgers.remove(segmentName);
+        deleteAllLedgers(ledger);
+    }
+
+    void deleteAllLedgers(LogStorage ledger) {
+        synchronized (ledger) {
```

Comment:
I'm not sure this is a great idea. The collection is threadsafe, and this now holds a lock over all the calls to deleteLedger which is a blocking operation.
I think there should be a threadsafe way to get a copy of the contents of the map. Perhaps drainTo?

Reply:
`threadsafe` ~= `synchronized`. Do you mean this leads to a deadlock?